common/utils.go

```diff
@@ -115,29 +113,60 @@ func hashWithSalt(password string, salt []byte, iterations int, algorithm string
 	buffer.WriteString(fmt.Sprintf("%s$", algorithm))
 	buffer.WriteString(strconv.Itoa(iterations))
 	buffer.WriteString("$")
-	buffer.WriteString(base64.StdEncoding.EncodeToString(salt))
+	// Re-encode salt, using encoding supplied in saltEncoding param
+	if saltEncoding == "utf-8" {
+		buffer.WriteString(string(salt))
+
+	} else {
+		buffer.WriteString(base64.StdEncoding.EncodeToString(salt))
+	}
 	buffer.WriteString("$")
 	buffer.WriteString(base64.StdEncoding.EncodeToString(hash))
-
+	//log.Debugf("Generated: ", buffer.String())
 	return buffer.String()
 }
 
 // HashCompare verifies that passed password hashes to the same value as the
 // passed passwordHash.
 // Taken from brocaar's lora-app-server: https://github.com/brocaar/lora-app-server
-func HashCompare(password string, passwordHash string) bool {
-	// SPlit the hash string into its parts.
+func HashCompare(password string, passwordHash string, saltEncoding string) bool {
+	//log.Debugf("Supplied:  ", passwordHash)
+	// Split the hash string into its parts.
 	hashSplit := strings.Split(passwordHash, "$")
+	// Check array is of expected length
 	if len(hashSplit) != 5 {
-		log.Error("invalid PBKDF2 hash supplied")
+		log.Errorf("HashCompare, invalid PBKDF2 hash supplied.")
 		return false
 	}
-
-	// Get the iterations and the salt and use them to encode the password
-	// being compared.cre
-	iterations, _ := strconv.Atoi(hashSplit[2])
-	salt, _ := base64.StdEncoding.DecodeString(hashSplit[3])
+	// Define empty error
+	var err error
+	// Get the iterations from PBKDF2 string
+	iterations, err := strconv.Atoi(hashSplit[2])
+	if err != nil {
+		log.Errorf("Error getting number of iterations from PBKDF2 hash.")
+		return false
+	}
+	// Convert salt to bytes, using encoding supplied in saltEncoding param
+	salt := []byte{}
+	if saltEncoding == "utf-8" {
+		salt = []byte(hashSplit[3])
+	} else {
```

Comment:
Again, switch on the valid encodings.

Reply:
Modified approach used in most recent commit.